### PR TITLE
Make clippy happy

### PR DESCRIPTION
**Pull request by trinity-1686a**
Clippy had many complains. I tried to fix those I could, and added explicit `allow(clippy::...)` for the others.

I left [`missing_safety_doc`](https://rust-lang.github.io/rust-clippy/master/#missing_safety_doc) alone: it triggers a lot, but in most case I'd lack the knowledge to actually do something intelligent about it.

### CODE REVIEW

File: src/main/network/router/mod.rs
Comment (stevenengler) [inline src/main/network/router/mod.rs:80]: We're in the middle of rewriting the `Router` and related networking modules, so I think we should avoid implementing `Default` until we know it doesn't require parameters to initialize.

File: src/main/utility/mod.rs
Comment (stevenengler) [inline src/main/utility/mod.rs:87]: This should be left as a warning. It's being tracked at https://github.com/shadow/shadow/issues/2514.

**Review by stevenengler — CHANGES_REQUESTED** [review]
Thanks for the PR!

File: src/lib/shmem/src/allocator.rs
Comment (trinity-1686a) [reply src/lib/shmem/src/allocator.rs:206]: the rational behind was that [clippy::inherent_to_string](https://rust-lang.github.io/rust-clippy/master/#inherent_to_string) was getting triggered. If you think I should instead revert and allow the lint, do tell

File: src/lib/shmem/src/scmutex.rs
Comment (trinity-1686a) [reply src/lib/shmem/src/scmutex.rs:126]: fwiw I'm not sure this code is sound, I think it needs an UnsafeCell

Comment (stevenengler) [reply src/lib/shmem/src/scmutex.rs]: The pointer does come from a `std::cell::UnsafeCell`, but it's a little convoluted because it needs a wrapper that is compatible with loom.

For `cfg(not(loom))`:

```rust
pub struct UnsafeCell<T>(std::cell::UnsafeCell<T>);
impl<T> UnsafeCell<T> {
    // [...]

    pub fn get_mut(&self) -> MutPtr<T> {
        MutPtr(self.0.get())
    }
}

pub struct MutPtr<T: ?Sized>(*mut T);
impl<T: ?Sized> MutPtr<T> {
    pub unsafe fn deref(&self) -> &mut T {
        unsafe { &mut *self.0 }
    }

    // [...]
}
```

Comment (trinity-1686a) [reply src/lib/shmem/src/scmutex.rs]: I believe (but may very well be wrong) upgrading a `&*mut T` to a `&mut T` isn't sound, even if the `*mut T` came from an `std::cell::UnsafeCell`. I think MutPtr needs to wrap a `*const std::cell::UnsafeCell<T>` and call [`raw_get`](https://doc.rust-lang.org/std/cell/struct.UnsafeCell.html#method.raw_get) (and transmute) on deref

Comment (trinity-1686a) [reply src/lib/shmem/src/scmutex.rs]: yeah, I'm now convinced you are right

**Comment by robgjansen** [comment]
@trinity-1686a thanks for the PR! Just FYI, we're going to wait for review from @sporksmith to help address some of the remaining items before moving forward.

**Review by stevenengler — APPROVED** [review]

File: src/lib/shmem/src/scmutex.rs
Comment (stevenengler) [reply src/lib/shmem/src/scmutex.rs]: Unresolving just so that @sporksmith can take a look.

File: src/main/host/syscall_types.rs
Comment (sporksmith) [reply src/main/host/syscall_types.rs:180]: I think I was just unfamiliar at the time with the convention of implementing `From` instead

**Review by sporksmith — CHANGES_REQUESTED** [review]

File: src/lib/shmem/src/scmutex.rs
Comment (sporksmith) [reply src/lib/shmem/src/scmutex.rs]: Ack; agreed that this should be sound and that suppressing the lint here makes sense. FWIW I believe we also exercise this path in unit tests run under miri.

**Comment by sporksmith** [comment]
Thanks for the cleanup! Once this gets merged maybe we should think about adding clippy to our lints GH workflow.

**Pull request merged**
